src/Nethermind/Nethermind.Era1/E2Store.cs

```diff
@@ -0,0 +1,281 @@
+// SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.Drawing;
+using System.IO.Compression;
+using System.Text;
+using DotNetty.Buffers;
+using Nethermind.Core.Crypto;
+using Newtonsoft.Json.Linq;
+using Snappier;
+using Snappy;
+
+namespace Nethermind.Era1;
+
+internal class E2Store : IDisposable
+{
+    internal const int HeaderSize = 8;
+    internal const int ValueSizeLimit = 1024 * 1024 * 50;
+
+    private Stream _stream;
+    private bool _disposedValue;
+
+    private EraMetadata? _metadata;
+    public long StreamLength => _stream.Length;
+    public EraMetadata Metadata => GetMetadata();
+
+    public E2Store(Stream stream)
+    {
+        _stream = stream;
+    }
+
+    private EraMetadata GetMetadata()
+    {
+        if (_metadata is null)
+        {
+            _metadata = ReadEraMetaData().GetAwaiter().GetResult();
+        }
+        return _metadata;
+    }
+
+    public async Task SetMetaData(CancellationToken token = default)
+    {
+        _metadata = await ReadEraMetaData(token);
+    }
+
+    public Task<int> WriteEntryAsSnappy(UInt16 type, byte[] bytes, CancellationToken cancellation = default)
+    {
+        return WriteEntry(type, bytes, true, cancellation);
+
+    }
+    public Task<int> WriteEntry(UInt16 type, byte[] bytes, CancellationToken cancellation = default)
+    {
+        return WriteEntry(type, bytes, false, cancellation);
+    }
+
+    private async Task<int> WriteEntry(UInt16 type, byte[] bytes, bool asSnappy, CancellationToken cancellation = default)
+    {
+        //TODO refactor to sync and use span?
+        byte[] headerBuffer = ArrayPool<byte>.Shared.Rent(HeaderSize);
```

Comment:
not sure if possible, but consider use `ArrayPoolList` for better readability

Reply:
changed to `ArrayPoolList `